Summary: pstat(8): display text for msts/endrun line discs
Forwarded: https://marc.info/?l=openbsd-tech&m=156777630806408

---

Index: src/usr.sbin/pstat/pstat.8
===================================================================
--- src.orig/usr.sbin/pstat/pstat.8
+++ src/usr.sbin/pstat/pstat.8
@@ -212,11 +212,17 @@ TTYDISC (see
 .Xr termios 4 ) ,
 .Ql ppp
 for PPPDISC (see
-.Xr ppp 4 )
-and
+.Xr ppp 4 ) ,
 .Ql nmea
 for NMEADISC (see
-.Xr nmea 4 ) .
+.Xr nmea 4 ) ,
+.Ql msts
+for MSTSDISC (see
+.Xr msts 4 ) ,
+and
+.Ql endrun
+for ENDRUNDISC (see
+.Xr endrun 4 ) .
 .El
 .It Fl v
 Print the active vnodes.
Index: src/usr.sbin/pstat/pstat.c
===================================================================
--- src.orig/usr.sbin/pstat/pstat.c
+++ src/usr.sbin/pstat/pstat.c
@@ -999,6 +999,12 @@ ttyprt(struct itty *tp)
 	case NMEADISC:
 		(void)printf("nmea\n");
 		break;
+	case MSTSDISC:
+		(void)printf("msts\n");
+		break;
+	case ENDRUNDISC:
+		(void)printf("endrun\n");
+		break;
 	default:
 		(void)printf("%d\n", tp->t_line);
 		break;
