From 4688daa012c0b839037bb5b10ba7f21122562623 Mon Sep 17 00:00:00 2001
From: "Iain R. Learmonth" <irl@fsfe.org>
Date: Sat, 23 Nov 2019 22:49:15 +0000
Subject: [PATCH 02/14] hambsd: modify the motd

---
 etc/motd | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/etc/motd b/etc/motd
index ff1dc9bb944..047662df624 100644
--- a/etc/motd
+++ b/etc/motd
@@ -1,10 +1,11 @@
-OpenBSD ?.? (UNKNOWN)
+HamBSD ?.? (UNKNOWN)
 
-Welcome to OpenBSD: The proactively secure Unix-like operating system.
+Welcome to HamBSD: The proactively secure Unix-like operating system for
+                   amateur radio.
 
-Please use the sendbug(1) utility to report bugs in the system.
-Before reporting a bug, please try to reproduce it with the latest
-version of the code.  With bug reports, please try to ensure that
-enough information to reproduce the problem is enclosed, and if a
-known fix for it exists, include that as well.
+Please do not use the sendbug(1) utility to report bugs in the system!
+Bugs can be reported directly to the developer: irl@fsfe.org.
+
+Please ensure you comply with local laws regarding amateur radio
+transmissions.
 
-- 
2.23.0

