From e369b4fb757352f524fcee9aa385c47fa725cc79 Mon Sep 17 00:00:00 2001
From: "Iain R. Learmonth" <irl@fsfe.org>
Date: Sat, 23 Nov 2019 23:17:42 +0000
Subject: [PATCH 07/13] ifconfig(8): show ax.25 addresses

---
 sbin/ifconfig/ifconfig.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sbin/ifconfig/ifconfig.c b/sbin/ifconfig/ifconfig.c
index 2ebb9b014cd..4071aee8f6c 100644
--- a/sbin/ifconfig/ifconfig.c
+++ b/sbin/ifconfig/ifconfig.c
@@ -69,6 +69,7 @@
 #include <net/if_media.h>
 #include <net/if_types.h>
 #include <netinet/in.h>
+#include <netax25/if_ax25.h>
 #include <netinet/in_var.h>
 #include <netinet6/in6_var.h>
 #include <netinet6/nd6.h>
@@ -3298,6 +3299,10 @@ status(int link, struct sockaddr_dl *sdl, int ls)
 	    (sdl->sdl_type == IFT_ETHER || sdl->sdl_type == IFT_CARP))
 		(void)printf("\tlladdr %s\n", ether_ntoa(
 		    (struct ether_addr *)LLADDR(sdl)));
+	if (sdl != NULL && sdl->sdl_alen &&
+	    (sdl->sdl_type == IFT_AX25))
+		(void)printf("\tlladdr %s\n", ax25_ntoa(
+		    (struct ax25_addr *)LLADDR(sdl)));
 
 	sep = '\t';
 #ifndef SMALL
-- 
2.23.0

