From 53bcbda707f812926e92f267254ccf02aed0e6de Mon Sep 17 00:00:00 2001
From: "Iain R. Learmonth" <irl@fsfe.org>
Date: Tue, 17 Dec 2019 21:08:16 +0000
Subject: [PATCH 08/20] ifconfig(8): show ax.25 addresses

---
 sbin/ifconfig/ifconfig.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sbin/ifconfig/ifconfig.c b/sbin/ifconfig/ifconfig.c
index 0ee441181c9..994f8053b7f 100644
--- a/sbin/ifconfig/ifconfig.c
+++ b/sbin/ifconfig/ifconfig.c
@@ -69,6 +69,7 @@
 #include <net/if_media.h>
 #include <net/if_types.h>
 #include <netinet/in.h>
+#include <netax25/if_ax25.h>
 #include <netinet/in_var.h>
 #include <netinet6/in6_var.h>
 #include <netinet6/nd6.h>
@@ -3300,6 +3301,10 @@ status(int link, struct sockaddr_dl *sdl, int ls)
 	    (sdl->sdl_type == IFT_ETHER || sdl->sdl_type == IFT_CARP))
 		(void)printf("\tlladdr %s\n", ether_ntoa(
 		    (struct ether_addr *)LLADDR(sdl)));
+	if (sdl != NULL && sdl->sdl_alen &&
+	    (sdl->sdl_type == IFT_AX25))
+		(void)printf("\tlladdr %s\n", ax25_ntoa(
+		    (struct ax25_addr *)LLADDR(sdl)));
 
 	sep = '\t';
 #ifndef SMALL
-- 
2.24.1

