Summary: pstat(8): display text for kiss line disc
---
Index: src/usr.sbin/pstat/pstat.8
===================================================================
--- src.orig/usr.sbin/pstat/pstat.8
+++ src/usr.sbin/pstat/pstat.8
@@ -219,10 +219,13 @@ for NMEADISC (see
 .Ql msts
 for MSTSDISC (see
 .Xr msts 4 ) ,
-and
 .Ql endrun
 for ENDRUNDISC (see
-.Xr endrun 4 ) .
+.Xr endrun 4 ) ,
+and
+.Ql kiss
+for KISSDISC (see
+.Xr kiss 4 ) .
 .El
 .It Fl v
 Print the active vnodes.
Index: src/usr.sbin/pstat/pstat.c
===================================================================
--- src.orig/usr.sbin/pstat/pstat.c
+++ src/usr.sbin/pstat/pstat.c
@@ -1005,6 +1005,9 @@ ttyprt(struct itty *tp)
 	case ENDRUNDISC:
 		(void)printf("endrun\n");
 		break;
+	case KISSDISC:
+		(void)printf("kiss\n");
+		break;
 	default:
 		(void)printf("%d\n", tp->t_line);
 		break;
