Summary: iclcd(8): identify compliant with license conditions
---
Index: src/usr.sbin/iclcd/Makefile
===================================================================
--- /dev/null
+++ src/usr.sbin/iclcd/Makefile
@@ -0,0 +1,9 @@
+#	$OpenBSD: Makefile,v 1.8 2015/03/11 14:59:04 deraadt Exp $
+
+PROG=	iclcd
+SRCS=	iclcd.c
+MAN=	iclcd.8
+
+LDADD=  -lpthread
+
+.include <bsd.prog.mk>
Index: src/usr.sbin/iclcd/iclcd.8
===================================================================
--- /dev/null
+++ src/usr.sbin/iclcd/iclcd.8
@@ -0,0 +1,56 @@
+.Dd September 7 2019
+.Dt ICLCD 8
+.Os
+.Sh NAME
+.Nm iclcd
+.Nd identification compliant with license conditions daemon
+.Sh SYNOPSIS
+.Nm
+.Op Fl D
+.Op Fl i Ar interface
+.Op Fl p Ar period
+.Sh DESCRIPTION
+.Nm
+sends
+.Em hamdex
+identification frames using the contents of
+.Pa /etc/mycallsign
+up to either the first newline or to the end of the file.
+.Pp
+.Nm
+is usually started at boot time, and can be enabled by
+setting the following in
+.Pa /etc/rc.conf.local :
+.Pp
+.Dl iclcd_flags=\&"\&"
+.Pp
+See
+.Xr rc 8
+and
+.Xr rc.conf 8
+for more information on the boot process
+and enabling daemons.
+.Sh OPTIONS
+.Bl -tag -width Ds
+.It Fl D
+Do not daemonize.
+.It Fl i Ar interface
+Interface name to broadcast ident frames (default: ekiss0).
+.It Fl p Ar period
+Period in seconds between broadcasts (default: 60).
+.Sh DISCLAIMER
+This may not be enough to comply with your local laws regarding identification
+of amateur radio transmissions. This manual page is not legal advice.
+.Sh CAVEATS
+It is not currently possible to broadcast identification on multiple interfaces
+with a single instance. It is however possible to manually start multiple
+instances.
+.Sh SEE ALSO
+.Xr hamdex 7 ,
+.Xr mycallsign 5 ,
+.Xr kiss 4
+.Sh AUTHORS
+The
+.Nm
+daemon was written for the HamBSD project by Iain R. Learmonth for the public
+domain.
Index: src/usr.sbin/iclcd/iclcd.c
===================================================================
--- /dev/null
+++ src/usr.sbin/iclcd/iclcd.c
@@ -0,0 +1,350 @@
+/*
+ * iclcd - intentification compliant with license conditions daemon
+ *
+ * Written by Iain R. Learmonth <irl@fsfe.org> for the public domain.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <syslog.h>
+#include <stdarg.h>
+#include <signal.h>
+#include <pthread.h>
+#include <errno.h>
+#include <sys/queue.h>
+
+#include <sys/times.h>
+#include <sys/time.h>
+#include <sys/ioctl.h>
+#include <net/bpf.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <netinet/if_ether.h>
+
+static struct bpf_insn insns[] = {
+	/* Check EtherType */
+	BPF_STMT(BPF_LD | BPF_H | BPF_ABS, 12),
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, 0x88B5, 0, 3),
+	/* Check HAMDEX SubType */
+	BPF_STMT(BPF_LD | BPF_B | BPF_ABS, 15),
+	BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, 0, 0, 1),
+	BPF_STMT(BPF_RET | BPF_K, (u_int)-1),
+	BPF_STMT(BPF_RET | BPF_K, 0),
+};
+
+static struct bpf_program filter = {
+	sizeof insns / sizeof(insns[0]),
+	insns
+};
+
+__dead void
+fatal(char* msg)
+{
+	syslog(LOG_DAEMON | LOG_EMERG,
+	    "iclcd hit fatal error; you might want to turn off your radio");
+	syslog(LOG_DAEMON | LOG_ERR,
+	    "%s", msg);
+	exit(1);
+}
+
+__dead void
+usage(void)
+{
+	extern char *__progname;
+
+	fprintf(stderr, "usage: %s [-D] [-i interface] [-p period]\n",
+	    __progname);
+	exit(1);
+}
+
+void
+signal_handler(int sig)
+{
+	switch(sig) {
+	case SIGHUP:
+		syslog(LOG_DAEMON | LOG_INFO, "caught hangup signal");
+		break;
+	case SIGTERM:
+		syslog(LOG_DAEMON | LOG_EMERG,
+		    "caught terminate signal, shutting down");
+		exit(0);
+		break;
+	}
+}
+
+char *
+mycallsign()
+{
+	FILE    *mcp;
+	char    *call, *nl;
+	size_t  callsize = 0;
+	ssize_t calllen;
+
+	call = NULL;
+
+	if ((mcp = fopen("/etc/mycallsign", "r")) == NULL)
+		fatal("could not open /etc/mycallsign");
+	if ((calllen = getline(&call, &callsize, mcp)) != -1) {
+		if ((nl = strchr(call, '\n')) != NULL)
+			nl[0] = '\0';
+		return call;
+	}
+	fatal("could not read callsign from /etc/mycallsign");
+}
+
+int
+composeframe(char *buf, char *call)
+{
+	int framelen, calllen;
+
+	const char header[] = {
+	  0xff, 0xff, 0xff, 0xff, 0xff, 0xff, // broadcast destination
+	  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // source filled by bpf
+	  0x88, 0xB5,                         // ethertype
+	  0x00,                               // subtype
+	  0x00                                // version
+	};
+
+	memcpy(buf, header, 16);
+	framelen = 16; // prefix is 15 bytes
+	calllen = strlcpy(&buf[16], call, 25);
+	if (calllen > 24) {
+		fatal("callsign in /etc/mycallsign too long");
+	}
+	framelen = framelen + calllen;
+	return framelen;
+}
+
+/*
+ * Open a BPF file and attach it to the interface named 'device'.
+ * Set immediate mode, check it is an Ethernet interface, and set a
+ * filter that only accepts iclc frames.
+ */
+int
+iclc_open(char *device)
+{
+	int bpf, immediate, iflen;
+	struct ifreq bound_if;
+	u_int dlt;
+
+	if ((bpf = open("/dev/bpf", O_RDWR)) == -1)
+		fatal("/dev/bpf failed to open");
+
+	/* Set immediate mode to process packets as they arrive. */
+	immediate = 1;
+	if (ioctl(bpf, BIOCIMMEDIATE, &immediate) == -1)
+		fatal("failed to set immediate mode");
+
+	/* Bind the network interface. */
+	iflen = strlen(device);
+	if (strlcpy(bound_if.ifr_name, device, sizeof(bound_if.ifr_name))
+	    < iflen)
+		fatal("interface name too long");
+	if(ioctl(bpf, BIOCSETIF, (caddr_t)&bound_if) == -1)
+		fatal("could not bind to interface");
+
+	/* Check the data link layer is Ethernet. */
+	if (ioctl(bpf, BIOCGDLT, (caddr_t)&dlt) == -1)
+		fatal("failed to get data link type");
+	if (dlt != DLT_EN10MB)
+		fatal("interface is not an ethernet");
+
+	/* Set filter program. */
+	if (ioctl(bpf, BIOCSETF, (caddr_t)&filter) == -1)
+		fatal("failed to set bpf filter");
+
+	return bpf;
+}
+
+void
+daemonize()
+{
+	int i;
+	// TODO: should this use daemon(3) instead? maybe it cuts this down
+	if (getppid() == 1)
+		return; /* already a daemon */
+	i = fork();
+	if (i<0)
+		exit(1); /* fork error */
+	if (i>0)
+		exit(0); /* parent exits */
+	/* child (daemon) continues */
+	setsid(); /* obtain a new process group */
+	for (i = getdtablesize(); i >= 0; --i)
+		close(i); /* close all descriptors */
+	i = open("/dev/null", O_RDWR);
+	dup(i);
+	dup(i); /* handle standard I/O */
+	signal(SIGCHLD, SIG_IGN); /* ignore child */
+	signal(SIGTSTP, SIG_IGN); /* ignore tty signals */
+	signal(SIGTTOU, SIG_IGN);
+	signal(SIGTTIN, SIG_IGN);
+	signal(SIGHUP, signal_handler); /* catch hangup signal */
+	signal(SIGTERM, signal_handler); /* catch kill signal */
+}
+
+struct iclc_beacon_attrs {
+	char	*call;   /* callsign as ascii text */
+	char	*device; /* interface name */
+	int	 bpf;    /* bpf file handle */
+	int	 period; /* beacon period */
+};
+
+void *
+iclc_beacon_loop(void *arguments) {
+	// TODO: make beaconing optional, we want this in the kernel
+	struct iclc_beacon_attrs *attrs = (struct iclc_beacon_attrs *)arguments;
+	char framebuf[41]; // extra byte for a terminating NUL, not txd
+	int framelen = composeframe(framebuf, attrs->call);
+	int unslept;
+
+	syslog(LOG_DAEMON | LOG_INFO,
+	    "started up beacon thread (interface %s, callsign: %s)", attrs->device, attrs->call);
+
+	for (;;) {
+		if (write(attrs->bpf, &framebuf, framelen) != framelen) {
+			syslog(LOG_DAEMON | LOG_EMERG,
+			    "failed to send ident frame, might want to unplug");
+			syslog(LOG_DAEMON | LOG_ERR,
+			    "failure reason: %m");
+		}
+		unslept = attrs->period;
+		while (unslept > 0)
+			unslept = sleep(unslept);
+	}
+}
+
+struct iclc_ident {
+	char			call[40]; // TODO: magic numbers are bad
+	struct ether_addr	hwaddr;
+	SLIST_ENTRY(iclc_ident)	entries;
+};
+SLIST_HEAD(iclc_idents_list, iclc_ident) iclc_idents;
+
+void
+iclc_write()
+{
+	// TODO: use a binary format instead (and write a tool that can query it)
+	struct iclc_ident *idp;
+	FILE *hh = fopen("/var/db/heards.txt", "w");
+	if (hh == NULL) {
+		fatal("error opening /var/db/heards.txt");
+	}
+	SLIST_FOREACH(idp, &iclc_idents, entries) {
+		fprintf(hh, "%02x:%02x:%02x:%02x:%02x:%02x\t%s\n",
+		    idp->hwaddr.ether_addr_octet[0],
+		    idp->hwaddr.ether_addr_octet[1],
+		    idp->hwaddr.ether_addr_octet[2],
+		    idp->hwaddr.ether_addr_octet[3],
+		    idp->hwaddr.ether_addr_octet[4],
+		    idp->hwaddr.ether_addr_octet[5],
+		    idp->call);
+	}
+	fclose(hh);
+}
+
+
+void
+iclc_process(u_char* buf)
+{
+	// TODO: deduplicate packets
+	// TODO: record a timestamp
+	struct ether_header *ep = (struct ether_header *)buf;
+	struct iclc_ident *idp = malloc(sizeof(struct iclc_ident));
+	syslog(LOG_DAEMON | LOG_INFO, "got callsign: %s from mac: %02x:%02x:%02x:%02x:%02x:%02x",
+	    (char *)(&buf[16]),
+	    buf[6],
+	    buf[7],
+	    buf[8],
+	    buf[9],
+	    buf[10],
+	    buf[11]);
+	memcpy(&idp->hwaddr, ep->ether_shost, 6);
+	strlcpy(idp->call, &buf[16], 40);
+	SLIST_INSERT_HEAD(&iclc_idents, idp, entries);
+	iclc_write();
+}
+
+int
+main(int argc, char **argv)
+{
+	int bpf, daemon, period;
+	char ch, *device;
+
+	/* option defaults */
+	daemon = 1;
+	period = 60;
+	device = "ekiss0"; // default
+
+	while ((ch = getopt(argc, argv, "i:p:D")) != -1) {
+		switch (ch) {
+		case 'i':
+			device = optarg;
+			break;
+		case 'p':
+			period = atoi(optarg);
+			break;
+		case 'D':
+			daemon = 0;
+			break;
+		default:
+			usage();
+			break;
+		}
+	}
+	argc -= optind;
+	argv += optind;
+
+	char *call = mycallsign();
+
+	/* Check for root privileges. */
+	if (geteuid())
+		fatal("need root privileges");
+
+	if (daemon)
+		daemonize();
+
+	/* This will not return if it fails. */
+	bpf = iclc_open(device);
+
+	struct iclc_beacon_attrs *beacon_attrs = malloc(sizeof(struct iclc_beacon_attrs));
+	beacon_attrs->call = call;
+	beacon_attrs->device = device;
+	beacon_attrs->bpf = bpf;
+	beacon_attrs->period = period;
+
+	pthread_t beacon_thread;
+	pthread_create(&beacon_thread, NULL, iclc_beacon_loop, beacon_attrs);
+
+	int rbufsize, cc;
+	u_char *rbuf, *bp, *ep;
+
+	/* Get buffer size needed for reads */
+	if (ioctl(bpf, BIOCGBLEN, (caddr_t)&rbufsize) == -1)
+		fatal("failed to get buffer size for bpf reads");
+	rbuf = malloc(rbufsize);
+	SLIST_INIT(&iclc_idents);
+
+	for (;;) {
+		cc = read(bpf, rbuf, rbufsize);
+		if (cc == -1 && errno == EINTR)
+			continue;
+		if (cc == -1)
+			fatal("read error");
+		bp = rbuf;
+		ep = bp + cc;
+		while (bp < ep) {
+#define caplen ((struct bpf_hdr *)bp)->bh_caplen
+#define hdrlen ((struct bpf_hdr *)bp)->bh_hdrlen
+			iclc_process(bp + hdrlen);
+			bp += BPF_WORDALIGN(hdrlen + caplen);
+		}
+	}
+
+	/* TODO: unreachable code, needs to go in signal handler? */
+	return pthread_join(beacon_thread, NULL);
+}
