Index: src/etc/rc.d/iclcd
===================================================================
--- /dev/null
+++ src/etc/rc.d/iclcd
@@ -0,0 +1,11 @@
+#!/bin/ksh
+#
+# $OpenBSD$
+
+daemon="/usr/sbin/iclcd"
+
+. /etc/rc.d/rc.subr
+
+rc_reload=NO
+
+rc_cmd $1
Index: src/etc/rc
===================================================================
--- src.orig/etc/rc
+++ src/etc/rc
@@ -584,7 +584,7 @@ start_daemon ldomd sshd switchd snmpd ld
 start_daemon relayd dhcpd dhcrelay mrouted dvmrpd radiusd eigrpd route6d
 start_daemon rad hostapd lpd smtpd slowcgi httpd ftpd
 start_daemon ftpproxy ftpproxy6 tftpd tftpproxy identd inetd rarpd bootparamd
-start_daemon rbootd mopd vmd spamd spamlogd sndiod
+start_daemon rbootd mopd vmd spamd spamlogd sndiod iclcd
 echo '.'
 
 # If rc.firsttime exists, run it just once, and make sure it is deleted.
Index: src/etc/rc.conf
===================================================================
--- src.orig/etc/rc.conf
+++ src/etc/rc.conf
@@ -27,6 +27,7 @@ ftpproxy6_flags=NO
 hostapd_flags=NO
 hotplugd_flags=NO
 httpd_flags=NO
+iclcd_flags=NO
 identd_flags=NO
 ifstated_flags=NO
 iked_flags=NO
Index: src/etc/Makefile
===================================================================
--- src.orig/etc/Makefile
+++ src/etc/Makefile
@@ -58,7 +58,7 @@ EXAMPLES_600=bgpd.conf doas.conf dvmrpd.
 
 # -r-xr-xr-x
 RCDAEMONS=amd apmd bgpd bootparamd cron dhcpd dhcrelay dhcrelay6 dvmrpd \
-	eigrpd ftpd ftpproxy ftpproxy6 hostapd hotplugd httpd identd ifstated \
+	eigrpd ftpd ftpproxy ftpproxy6 hostapd hotplugd httpd iclcd identd ifstated \
 	iked inetd isakmpd iscsid ldapd ldattach ldomd ldpd lockd \
 	lpd mopd mountd mrouted nfsd npppd nsd ntpd ospf6d ospfd \
 	pflogd portmap rad radiusd rarpd rbootd rebound relayd ripd route6d \
