Index: src/distrib/sets/lists/base/mi
===================================================================
--- src.orig/distrib/sets/lists/base/mi
+++ src/distrib/sets/lists/base/mi
@@ -154,6 +154,7 @@
 ./etc/rc.d/hostapd
 ./etc/rc.d/hotplugd
 ./etc/rc.d/httpd
+./etc/rc.d/iclcd
 ./etc/rc.d/identd
 ./etc/rc.d/ifstated
 ./etc/rc.d/iked
@@ -2512,6 +2513,7 @@
 ./usr/sbin/host
 ./usr/sbin/hostapd
 ./usr/sbin/httpd
+./usr/sbin/iclcd
 ./usr/sbin/identd
 ./usr/sbin/ifstated
 ./usr/sbin/ikectl
Index: src/distrib/sets/lists/man/mi
===================================================================
--- src.orig/distrib/sets/lists/man/mi
+++ src/distrib/sets/lists/man/mi
@@ -2322,6 +2322,7 @@
 ./usr/share/man/man8/i386/memconfig.8
 ./usr/share/man/man8/i386/pxeboot.8
 ./usr/share/man/man8/i386/wsmoused.8
+./usr/share/man/man8/iclcd.8
 ./usr/share/man/man8/identd.8
 ./usr/share/man/man8/ifconfig.8
 ./usr/share/man/man8/ifstated.8
